### PR TITLE
IridescentDishWithOlives for ior, specular and volume extensions

### PR DESCRIPTION
**Pull request by echadwick-wayfair**
A glass dish with olives and matching glass cover. This is a “mascot” model for Phase 2 extensions, showcasing specular color, ior and volume extensions.  An animation has been included to lift the refractive glass cover.

glassCover_animation

Edit: updating the animated GIF to match the latest commit.

### CODE REVIEW

File: 2.0/PearlGoldOliveCoveredGlassDish/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# PearlGoldOliveCoveredGlassDish 
+
+## Screenshots
+
+![screenshot Babylon.js](screenshot/screenshot_large.jpg)
```
Comment (lexaknyazev) [inline 2.0/PearlGoldOliveCoveredGlassDish/README.md:5]: The file is called `screenshot_Large.jpg` (with capital L) so it doesn't show up on case-sensitive filesystems, including GitHub preview. Please either rename it or fix the link.

Comment (echadwick-wayfair) [reply 2.0/PearlGoldOliveCoveredGlassDish/README.md:5]: Should all the screenshots be renamed using lowercase? Or just screenshot_Large.jpg?

Comment (lexaknyazev) [reply 2.0/PearlGoldOliveCoveredGlassDish/README.md:5]: Just `screenshot_Large.jpg`. There are many mixed-case files in the repo.

**Comment by cx20** [comment]
I experimentally added this model to [gltf-test](https://cx20.github.io/gltf-test/).
The current display results are as follows.

|Library|result|
|:----:|:----:|
|[Three.js r130](https://cx20.github.io/gltf-test/examples/threejs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|image|
|[Babylon.js v5.0.0 alpha.30](https://cx20.github.io/gltf-test/examples/babylonjs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|image|
|[Filament v1.10.5](https://cx20.github.io/gltf-test/examples/filament/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|image|

**Comment by echadwick-wayfair** [comment]
Thanks to both of you for checking the submission. 

Looks like I need to adjust the vertex normals on the gold leaf. The outer rim is looking at bit donut-shaped, rather than flat.

Filament seems a bit aggressive with transmission roughness, not sure what I can do there.

**Comment by cx20** [comment]
@echadwick-wayfair 

`HDRI Haven`(hdrihaven.com) has recently moved to `Poly Haven`(polyhaven.com), so I think you should change the link.
before) https://hdrihaven.com/hdri/?h=artist_workshop
after) https://polyhaven.com/a/artist_workshop

**Comment by cx20** [comment]
@echadwick-wayfair Thank you for the quick fix.

And if possible, To be consistent with other samples, it is better to name the sample model file the same as the folder name is a good idea.

|Folder Name     |File Name|
|:---------------|:--------|
|`Duck`/glTF       |`Duck`.gltf|
|`Duck`/glTF-Binary|`Duck`.glb |

|Folder Name     |File Name|
|:----|:-----|
|`PearlGoldOliveCoveredGlassDish`/glTF|`PearlGoldOliveCoveredGlassDish`.gltf|
|`PearlGoldOliveCoveredGlassDish`/glTF-Pathtracer|`PearlGoldOliveCoveredGlassDish`.gltf|

**Comment by cx20** [comment]
Thanks for the update.
As for [gltf-test](https://cx20.github.io/gltf-test/), I have also updated the model and added a link to `Pathtracer`. After some time, it will be reflected in GitHub-Pages.

image

**Comment by cx20** [comment]
For reference, the current display results are as follows.

|model|Three.js r130|Babylon.js v5.0.0 alpha.30|Filament v1.10.5|
|:-----:|:-----:|:-----:|:-----:|
|[glTF](https://github.com/KhronosGroup/glTF-Sample-Models/tree/PearlGoldOliveCoveredGlassDish/2.0/PearlGoldOliveCoveredGlassDish/glTF)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/threejs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/babylonjs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/filament/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF)|
|[Pathtracer](https://github.com/KhronosGroup/glTF-Sample-Models/tree/PearlGoldOliveCoveredGlassDish/2.0/PearlGoldOliveCoveredGlassDish/glTF-Pathtracer)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/threejs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF-Pathtracer)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/babylonjs/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF-Pathtracer)|[image](https://rawcdn.githack.com/cx20/gltf-test/a4b8369ce788c44dc8fcf2e68b9c17514e13283f/examples/filament/index.html?category=tutorialModels&model=PearlGoldOliveCoveredGlassDish&scale=8&type=glTF-Pathtracer)|

@romainguy Are the results displayed by Filament as expected?

**Comment by romainguy** [comment]
Volume and specular are not implemented by Filament. Where can I download the glTF?

**Comment by cx20** [comment]
@romainguy Thank you for your confirmation. The model is below.
https://github.com/KhronosGroup/glTF-Sample-Models/tree/PearlGoldOliveCoveredGlassDish/2.0/PearlGoldOliveCoveredGlassDish/glTF
https://github.com/KhronosGroup/glTF-Sample-Models/tree/PearlGoldOliveCoveredGlassDish/2.0/PearlGoldOliveCoveredGlassDish/glTF-Pathtracer

**Comment by echadwick-wayfair** [comment]
The version within glTF-Pathtracer is meant for raytracers, it does not appear correctly in rasterizers (roughness too low, etc., see readme for details). The three renderers you're showing here are only rasterizers I think?

**Comment by romainguy** [comment]
Testing locally it seems like Filament has an issue with roughness mapping.

**Comment by abwood** [comment]
> Testing locally it seems like Filament has an issue with roughness mapping.

The [TransmissionRoughnessTest](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/TransmissionRoughnessTest) sample model may be helpful here.

**Comment by romainguy** [comment]
When using screen space refraction/transmission, Filament attempts to remap the roughness to try and match the screen space effect's gaussian approximation of the GGX NDF. It's that remapping that increases roughness too much it seems.

**Comment by romainguy** [comment]
Is there a screenshot of the path traced reference with the same roughness (0.15 instead of 0.035) used in the rasterizer asset?

**Comment by romainguy** [comment]
I'm actually starting to doubt three.js and Babylon here. I would expect the result to be more blurry with a roughness of 0.15. I think Filament blurs too much but three/Babylon too little. I did change Filament's roughness remapping to match the glTF spec (at IOR 1.0, roughness shouldn't apply to transmission) but otherwise our screen space refraction matches cubemap refraction at equivalent roughness.

According to @sebavan, Babylon only uses regular texture mip levels for screen space refraction where Filament uses roughness-aware Gaussian-filtered mip levels, which would explain the difference.

**Comment by sebavan** [comment]
Yup I can validate that the blur would be more accurate from filament than babylon in this case as we chose to only optim for perf in this case.

**Comment by echadwick-wayfair** [comment]
If equivalent roughness cannot be rendered at a performant level on low- or mid-level devices (e.g. mobile) then do we still need different materials for high-end devices and non-realtime renderers?

**Comment by echadwick-wayfair** [comment]
What about the issue of crossing multiple boundaries in raytracers, vs. a single boundary in rasterizers? 

If a glass surface is set to "roughnessFactor":0.15 then wouldn't this cause a raytracer to apply a total roughness of 0.6?

Pathtracer_Roughness

**Comment by sebavan** [comment]
The material can actually be the same, this is only the refraction texture creation and sampling which uses different techniques.

About this issue, I think @MiiBond had implemented smthg for it in Babylon (simulation of course) and could probably detail more ?

**Comment by cx20** [comment]
@echadwick-wayfair It's a trivial matter, but the title of `README.md` is still old (_PearlGoldOliveCoveredGlassDish_).

https://github.com/KhronosGroup/glTF-Sample-Models/blob/PearlGoldOliveCoveredGlassDish/2.0/IridescentDishWithOlives/README.md#pearlgoldolivecoveredglassdish

**Comment by donmccurdy** [comment]
Of the four extensions (KHR_materials_ior, KHR_materials_specular, KHR_materials_transmission, KHR_materials_volume) three.js currently implements only transmission, so I'm afraid it won't be much help for comparison here.

**Comment by romainguy** [comment]
Speaking of which, I just added KHR_materials_volume to Filament. It will be in 1.11.0 which we'll release on July 19. It can be tested in our main branch right now.

**Comment by coderextreme** [comment]
https://playcanv.as/p/wQgQBgkE/

Looky, real-time animated 3 channel chromatic dispersion (even runs on some smartphones).   Thanks to J3D, mick and Web3d folks!

**Comment by donmccurdy** [comment]
@coderextreme the sample here is for supporting particular definitions of PBR Transmission, IOR, and Volume defined by very recent [extensions to the glTF format](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos)... it doesn't look like the [PlayCanvas glTF implementation](https://playcanvas.com/viewer) supports those just yet (screenshot below), but it would be an awesome addition if so. 🙂

Screen Shot 2021-07-09 at 11 41 39 PM

**Comment by cx20** [comment]
I updated the library and model and tried to display it again.

|Library|IridescentDishWithOlives.gltf - closed|IridescentDishWithOlives.gltf - open|
|:-----:|:-----:|:-----:|
|[glTF Sample Viewer](https://github.khronos.org/glTF-Sample-Viewer-Release/)|image|image|
|[Three.js r131dev + PR](http://localhost:3000/examples/threejs/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|image|
|[Babylon.js v5.0.0 alpha.30](http://localhost:3000/examples/babylonjs/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|image|
|[Filament v1.11.0rc](http://localhost:3000/examples/filament/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|image|

**Comment by donmccurdy** [comment]
Thanks to @takahirox's work, three.js will support 3 of the 4 extensions soon (just missing `KHR_materials_specular`). I assume that is the reason the iridescent effect is missing in the three.js render above.

**Comment by emackey** [comment]
Any further discussion needed on this model?  I think this looks good to merge.

**Comment by echadwick-wayfair** [comment]
There's a bit more I could add to the readme to explain why the roughness is different on the inside of the glassCover, and how the specularColor is being used. But I think the asset itself is complete.

**Comment by emackey** [comment]
Looks like the [press release is hitting the wire now](https://www.khronos.org/news/press/new-gltf-extensions-raise-the-bar-on-3d-asset-visual-realism).  🚀

**Comment by cx20** [comment]
I updated the development version of `Three.js` and tried testing this model again.
Perhaps the `thicknessTexture` is now handled correctly and I think it has improved.

|Library|IridescentDishWithOlives.gltf - open|
|:-----:|:-----:|
|[Three.js r131dev (3 days ago)](https://rawcdn.githack.com/cx20/gltf-test/1b1d1c76bb40a966dcf888430a2e064dbc953bfb/examples/threejs/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|
|[Three.js r131dev (latest)](https://rawcdn.githack.com/cx20/gltf-test/98cffbf81a3d2feba98d4ec89c4c1544ff7918af/examples/threejs/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|

BTW, it is a [known problem](https://github.com/cx20/gltf-test/issues/68#issuecomment-757134470), but there is a problem with CubeMap/IBL used in the `Three.js` sample of [gltf-test](https://github.com/cx20/gltf-test).
For example, the left and right are reversed.

The CubeMap/IBL you are currently using refers to an [older version](https://github.com/KhronosGroup/glTF-Sample-Environments/tree/4eace30f795fa77f6e059e3b31aa640c08a82133/papermill/specular) in the `glTF-Sample-Environments` repository. I think it has been fixed in the [latest version](https://github.com/KhronosGroup/glTF-Sample-Environments/blob/master/papermill/ggx/specular.ktx2), but it is in `ktx2` format and cannot be used as CubeMap/IBL as it is. And because it uses LFS, it can't be hosted on GitHub Pages.
Anyone who can solve this problem is welcome.

**Comment by cx20** [comment]
> BTW, it is a [known problem](https://github.com/cx20/gltf-test/issues/68#issuecomment-757134470), but there is a problem with CubeMap/IBL used in the `Three.js` sample of [gltf-test](https://github.com/cx20/gltf-test).
> For example, the left and right are reversed.

I think that the problem of CubeMap/IBL of the Three.js sample of gltf-test has been solved.
Thanks to the [Rhodonite team](https://github.com/actnwit/RhodoniteTS) for working on improving the cubemap texture.

|Library|IridescentDishWithOlives.gltf - closed|IridescentDishWithOlives.gltf - open|
|:-----:|:-----:|:-----:|
|[glTF Sample Viewer](https://github.khronos.org/glTF-Sample-Viewer-Release/)|image|image|
|[Three.js r131dev](https://cx20.github.io/gltf-test/examples/threejs/index.html?category=tutorialModels&model=IridescentDishWithOlives&scale=8&type=glTF)|image|image|

**Comment by cx20** [comment]
@echadwick-wayfair As a minor question, is it intentional that relative paths are used to reference textures in the current directory?
https://github.com/KhronosGroup/glTF-Sample-Models/blob/PearlGoldOliveCoveredGlassDish/2.0/IridescentDishWithOlives/glTF/IridescentDishWithOlives.gltf#L613-L645
```json
"uri": "./glassdish_spec.png"
```
The glTF specification does not have a problem with relative paths being used, but I have tried several online glTF viewers (For example, PlayCanvas glTF Viewer) and they seem to give [an error](https://forum.playcanvas.com/t/the-viewer-will-fail-if-the-gltf-model-contains-relative-paths/20976) if relative paths are used. (This is probably a problem that needs to be fixed by the glTF viewer.

However, I don't think it is necessary to specify a relative path because the sample model should be simple.

**Comment by echadwick-wayfair** [comment]
This is an artifact from the Max2Babylon exporter. Thanks for the info about compatibility. I'll file an Issue in their repo, and remove them from this model as I think you're right, these seem to be redundant here.

**Comment by echadwick-wayfair** [comment]
Max2Babylon issue reported here https://github.com/BabylonJS/Exporters/issues/985

**Review by emackey — APPROVED** [review]
Thanks @echadwick-wayfair!

**Pull request merged**